src/compiler/transformers/es2020.ts

```diff
@@ -90,27 +104,27 @@ namespace ts {
             return thisArg ? factory.createSyntheticReferenceExpression(expression, thisArg) : expression;
         }
 
-        function visitNonOptionalCallExpression(node: CallExpression, captureThisArg: boolean): Expression {
+        function visitNonOptionalCallExpression(node: CallExpression, captureThisArg: boolean, isExpressionStatement: boolean): Expression {
             if (isOptionalChain(node)) {
                 // If `node` is an optional chain, then it is the outermost chain of an optional expression.
-                return visitOptionalExpression(node, captureThisArg, /*isDelete*/ false);
+                return visitOptionalExpression(node, captureThisArg, /*isDelete*/ false, /*hasUnusedOrFalsyResult*/ isExpressionStatement);
             }
             return visitEachChild(node, visitor, context);
         }
 
-        function visitNonOptionalExpression(node: Expression, captureThisArg: boolean, isDelete: boolean): Expression {
+        function visitNonOptionalExpression(node: Expression, captureThisArg: boolean, isDelete: boolean, isExpressionStatement: boolean): Expression {
             switch (node.kind) {
-                case SyntaxKind.ParenthesizedExpression: return visitNonOptionalParenthesizedExpression(node as ParenthesizedExpression, captureThisArg, isDelete);
+                case SyntaxKind.ParenthesizedExpression: return visitNonOptionalParenthesizedExpression(node as ParenthesizedExpression, captureThisArg, isDelete, isExpressionStatement);
                 case SyntaxKind.PropertyAccessExpression:
-                case SyntaxKind.ElementAccessExpression: return visitNonOptionalPropertyOrElementAccessExpression(node as AccessExpression, captureThisArg, isDelete);
-                case SyntaxKind.CallExpression: return visitNonOptionalCallExpression(node as CallExpression, captureThisArg);
+                case SyntaxKind.ElementAccessExpression: return visitNonOptionalPropertyOrElementAccessExpression(node as AccessExpression, captureThisArg, isDelete, isExpressionStatement);
+                case SyntaxKind.CallExpression: return visitNonOptionalCallExpression(node as CallExpression, captureThisArg, isExpressionStatement);
                 default: return visitNode(node, visitor, isExpression);
             }
         }
 
-        function visitOptionalExpression(node: OptionalChain, captureThisArg: boolean, isDelete: boolean): Expression {
+        function visitOptionalExpression(node: OptionalChain, captureThisArg: boolean, isDelete: boolean, hasUnusedOrFalsyResult: boolean): Expression {
             const { expression, chain } = flattenChain(node);
-            const left = visitNonOptionalExpression(expression, isCallChain(chain[0]), /*isDelete*/ false);
+            const left = visitNonOptionalExpression(expression, isCallChain(chain[0]), /*isDelete*/ false, hasUnusedOrFalsyResult);

```

Comment:
`hasUnusedOrFalsyResult` should probably be changed to `isExpressionStatement` because of this call. I wanted to avoid having individual args for every optimizable descendents case as this visit only cares if it has an unused result. Beyond the expression statement case we can start using a single extra arg but now I'm not sure thats the best idea.